frontend/ui/widget/menu.lua

```diff
@@ -281,34 +286,116 @@ function MenuItem:init()
+            self.font_size = self.font_size - 1
+            item_name = TextBoxWidget:new {
+                text = self.text,
+                face = Font:getFace(self.font, self.font_size),
+                width = self.content_width - mandatory_w - state_button_width - text_mandatory_padding,
+                alignment = "left",
+                bold = self.bold,
+                fgcolor = self.dim and Blitbuffer.COLOR_DARK_GRAY or nil,
+            }
+            local height = item_name:getSize().h
+            if height < max_item_height then -- we fit !
+                offset = #item_name.charlist
+                break
+            end
+
+            -- if text is too height with that really small font we don't show text at all
+            -- this shouldn't happen
+            if self.font_size <= 8 then
+                item_name = TextBoxWidget:new {
+                    text = "…",
+                    face = Font:getFace(self.font, self.font_size),
+                    width = self.content_width - mandatory_w - state_button_width - text_mandatory_padding,
+                    alignment = "left",
+                    bold = self.bold,
+                    fgcolor = self.dim and Blitbuffer.COLOR_DARK_GRAY or nil,
+                }
+                flag_add_ellipsis = false
+                break
+            end
+        end
+        -- add ellipsis when text was truncated
+        if flag_add_ellipsis then
+            local text_last_line
+            -- when lines is more than 1 we see only for last visible line
+            if num_lines > 1 then
+                local offset_prev = item_name_orig.vertical_string_list[num_lines - 1].offset
+                text_last_line = table.concat(item_name_orig.charlist, '', offset_prev, offset)
+            else
+                text_last_line = self.text
+            end
+            local text_size = RenderText:sizeUtf8Text(0, self.content_width,
+                Font:getFace(self.font, self.font_size), text_last_line, true, self.bold).x
+            local ellipsis_size = RenderText:sizeUtf8Text(0, self.content_width,
+                Font:getFace(self.font, self.font_size), "…", true, self.bold).x
+
+            local text_size_increase = text_size
+            local max_offset = #item_name_orig.charlist
+            -- try to add chars to better align
+            while item_name.width > text_size_increase + ellipsis_size and offset < max_offset do
+                text_size_increase = text_size_increase + item_name_orig:getCharWidth(offset + 1)
+                if text_size_increase + ellipsis_size < item_name.width then
+                    -- add one char to text
+                    offset = offset + 1
                 end
-                self.text = table.concat(item_name.charlist, '', 1, offset) .. "…"
-                flag_fit = true
+            end
+            -- remove chars when text is too long
+            while item_name.width <= text_size + ellipsis_size do
```

Comment:
And the, why does it happen you need to remove chars? (because you added some and made sure they fitted with the ellipsis)?
Or does this happen and is only needed when you actually did not have to add some chars? So, just to get rid of the orignal chars on that last line, to make room for the ellipsis?

Reply:
Second sentence. It's to make a room for ellipsis.